test/test_trio_server/mock_lean.py

```diff
@@ -0,0 +1,152 @@
+        self.messages: deque[Dict] = deque()
+        self.partial_message: bytes = b""
+        self.script: List[LeanScriptStep] = script
+
+    def kill(self):
+        pass
+
+    @staticmethod
+    def parse_message(b: bytes) -> Dict:
+        return json.loads(b.decode())
+
+    @staticmethod
+    def messages_are_equal(m1: Dict, m2: Dict):
+        s1 = json.dumps(m1, sort_keys=True)
+        s2 = json.dumps(m2, sort_keys=True)
+        return s1 == s2
+
+    def collect_stdin_messages(self):
+        try:
+            data = self.stdin.get_data_nowait()
+        except trio.WouldBlock:  # no data
+            return
+
+        raw_messages = (self.partial_message + data).split(b"\n")
+        self.partial_message = raw_messages.pop()
+        self.messages.extend(self.parse_message(m) for m in raw_messages)
+
+    async def assert_message_is_received(self, message_expected):
+        await trio.sleep(0.01)
+        self.collect_stdin_messages()
+
+        assert self.messages, f"Mock Lean was expecting\n{message_expected}\nbut no messages were received."
+
+        message_received = self.messages.pop()
+        assert self.messages_are_equal(message_expected, message_received), \
+            f"Mock Lean was expecting\n{message_expected}\nbut received\n{message_received}"
+
+    async def assert_no_messages_received(self):
+        await trio.sleep(0.01)
+        self.collect_stdin_messages()
+
+        assert not self.messages, f"Mock Lean was not expecting as message, but \n{self.messages[0]}\nwas received."
+
+    def send_bytes(self, message_bytes: bytes):
+        self.stdout.put_data(message_bytes)
+
+    def send_message(self, message):
+        message_bytes = json.dumps(message).encode() + b"\n"
+        self.stdout.put_data(message_bytes)
+
+    async def follow_script(self):
+        for step in self.script:
+            if isinstance(step, LeanTakesTime):
+                print(f"\nLean is taking {step.seconds} seconds before doing anything.")
+                await trio.sleep(step.seconds)
+            elif isinstance(step, LeanShouldGetRequest):
+                print(f"\nLean should receive the following request:\n{step.message}")
+                await self.assert_message_is_received(step.message)
+            elif isinstance(step, LeanSendsResponse):
+                print(f"\nLean sends the following response:\n{step.message}")
+                self.send_message(step.message)
+            elif isinstance(step, LeanSendsBytes):
+                print(f"\nLean sends the following bytes:\n{step.message_bytes}")
+                self.send_bytes(step.message_bytes)
+            elif isinstance(step, LeanShouldNotGetRequest):
+                print(f"\nLean should not have received any requests yet.")
+                await self.assert_no_messages_received()
+
+
+async def start_with_mock_lean(lean_server: TrioLeanServer, script: List[LeanScriptStep]):
```

Comment:
You could also monkeypatch `TrioLeanServer` in your test, but let's see if your way is convenient enough.